### DISCUSSION TITLE
EntityIterable cache doesn't appear to respect ENTITY_ITERABLE_CACHE_SIZE

### DISCUSSION THREAD

**Discussion by gravelld** [JetBrains/xodus#112] [Q&A]
I'm doing this:

        PersistentEntityStoreConfig config = new PersistentEntityStoreConfig();
        config = config.setEntityIterableCacheSize(100);
        this.entityStore = PersistentEntityStores.newInstance(config, env, getClass().getSimpleName());

I then load some data into my Xodus entity store and take a heap dump at the end.

No matter if I `setEntityIterableCacheSize` or leave it at the default the end memory consumption is the same - about 15MB. The same structure is built by the EntityIterableCache - a `PersistentLinkedHashMap` with the same number of levels and objects inside. I can see `NonAdjustablePersistentObjectCache` has the size correctly passed through.

I can see there's a `RemoveEldestFunction` which appears to do cache eviction. It asks for the size of the contained map, and this is implemented by `RootTableNode` which currently has its size set to 5 - thus it never evicts.

Do I misunderstand the purpose of this setting? I guess so.

(Aside)
`MAX_DEFAULT_ENTITY_ITERABLE_CACHE_SIZE` has an odd name - it appears to not be a maximum but a minimum:

        return Math.max((int) (Runtime.getRuntime().maxMemory() >> 20), MAX_DEFAULT_ENTITY_ITERABLE_CACHE_SIZE);

(i.e. a number related to the heap size _or_ `MAX_DEFAULT_ENTITY_ITERABLE_CACHE_SIZE`, whichever is larger)

**Comment by andrii0lomakin** [comment]
Hi @gravelld you misunderstood object which you should observe, it should be `ConcurrentObjectCache` instance. Which field do you mean when you mention `PersistentLinkedHashMap` ?

**Comment by gravelld** [comment]
I saw `ConcurrentLongObjectCache`  (note the additional "`Long`") but I thought that was the Log cache, not the EntityIterableCache. I thought these were two separate things. The `ConcurrentLongObjectCache` I see is stored within `SharedLogCache`.

The `EntityIterableCache` I'm referring to is stored in the `iterableCache` member of `PersistentEntityStoreImpl`. The `EntityIterableCache` then has a `cache -> root` and then there's a `Root` instance which constains the `Persistentx` hashmaps/hashsets.

This iterable cache always weighs in at a hefty 12MB so I was hoping to reduce its size (I'm running in a constrained environment). And besides - I thought the `setEntityIterableCacheSize` referred to a cache of EntityIterables and was nothing to do with the Log.

The Log caching sizing _does_ work, btw. For that I use:

        EnvironmentConfig envConfig = new EnvironmentConfig();
        envConfig = envConfig.setMemoryUsage((1024 * 1024) * 2);

**Comment by andrii0lomakin** [comment]
@gravelld classes you list are part of the query processing, not the cache itself. If you wish us to investigate this issue, please fill the case in the bug tracker and provide a reproducer. As for now, it is hard to understand if the problem does exist. We use the given parameter daily 
 on many installations, so we need to double-check that the issue is present.

**Comment by gravelld** [comment]
Thanks - I'll revisit the issue assuming it's a query that is being processed or used.
